File: js/src/algo_gen/tier.js
/**
 * tier.js — Composite tier scorer for the S1 algorithm generator.
 *
 * The tier is the MINIMUM of the three component scores.
 * Mirrors python/algoforge/algo_gen/tier.py exactly.
 *
 * Public API:
 *   scoreToTier(walkForward, mc, robustness, cfg?) -> TierReport
 */

import { TierName, makeTierReport } from './types.js';

// ── Hard-coded defaults (mirrors Python _DEFAULT_CFG) ─────────────────────────

const DEFAULT_CFG = {
    orange_size_mult:              0.25,
    yellow_size_mult:              0.50,
    green_size_mult:               1.00,
    white_size_mult:               1.50,
    require_manual_review_below:   'green',
    paper_only_below:              'yellow',
};

// Tier thresholds (min_score → tier), sorted descending
const TIER_THRESHOLDS = [
    { threshold: 95.0, tier: TierName.white },
    { threshold: 90.0, tier: TierName.green },
    { threshold: 80.0, tier: TierName.yellow },
    { threshold: 70.0, tier: TierName.orange },
];
// Below 70 → red

/**
 * Merge user config overrides into defaults.
 * @param {object|null} cfg
 * @returns {object}
 */
function loadCfg(cfg) {
    const resolved = { ...DEFAULT_CFG };
    if (cfg) Object.assign(resolved, cfg);
    return resolved;
}

/**
 * Determine tier from min score.
 * @param {number} minScore
 * @returns {string}  TierName value
 */
export function tierFromMinScore(minScore) {
    for (const { threshold, tier } of TIER_THRESHOLDS) {
        if (minScore >= threshold) return tier;
    }
    return TierName.red;
}

/**
 * Assign a tier from three component scores.
 *
 * @param {number}      walkForward
 * @param {number}      mc
 * @param {number}      robustness
 * @param {object|null} [cfg]  optional overrides
 * @returns {import('./types.js').TierReport}
 */
export function scoreToTier(walkForward, mc, robustness, cfg) {
    const loadedCfg = loadCfg(cfg || null);
    const minScore  = Math.min(walkForward, mc, robustness);
    const tier      = tierFromMinScore(minScore);

    // Resolve size multiplier
    const sizeMult = {
        [TierName.red]:    0.0,
        [TierName.orange]: Number(loadedCfg.orange_size_mult),
        [TierName.yellow]: Number(loadedCfg.yellow_size_mult),
        [TierName.green]:  Number(loadedCfg.green_size_mult),
        [TierName.white]:  Number(loadedCfg.white_size_mult),
    }[tier];

    // Experimental flag: orange and yellow (and red)
    const experimental = tier === TierName.red || tier === TierName.orange || tier === TierName.yellow;

    // Paper-only flag: depends on paper_only_below config
    const paperOnlyBelow = loadedCfg.paper_only_below;
    let paperOnly;
    if (paperOnlyBelow === 'yellow') {
        paperOnly = tier === TierName.red || tier === TierName.orange;
    } else if (paperOnlyBelow === 'green') {
        paperOnly = tier === TierName.red || tier === TierName.orange || tier === TierName.yellow;
    } else {
        paperOnly = tier === TierName.red || tier === TierName.orange;
    }

    // Round to 4 decimal places (mirrors Python round(x, 4))
    const round4 = x => Math.round(x * 10000) / 10000;

    return makeTierReport({
        tier,
        minScore:    round4(minScore),
        walkForward: round4(walkForward),
        mcBootstrap: round4(mc),
        robustness:  round4(robustness),
        sizeMult,
        experimental,
        paperOnly,
        stages: [],
    });
}

File: cpp/src/algo_gen/tier.cpp
/**
 * AlgoForge — src/algo_gen/tier.cpp
 * S1 Phase 1 — Composite tier scorer.
 *
 * Mirrors python/algoforge/algo_gen/tier.py:
 *   tier = f(min(walk_forward, mc, robustness))
 *   thresholds: white>=95, green>=90, yellow>=80, orange>=70, else red
 */

#include "core/algo_gen.hpp"

namespace algoforge::algo_gen {

/* =========================================================================
 * Tier enum helpers
 * ========================================================================= */

const char* tier_name(Tier t) noexcept {
    switch (t) {
        case Tier::Red:    return "red";
        case Tier::Orange: return "orange";
        case Tier::Yellow: return "yellow";
        case Tier::Green:  return "green";
        case Tier::White:  return "white";
    }
    return "red";
}

const char* tier_color(Tier t) noexcept {
    switch (t) {
        case Tier::Red:    return "#ff4444";
        case Tier::Orange: return "#ff8800";
        case Tier::Yellow: return "#ffdd00";
        case Tier::Green:  return "#00ff88";
        case Tier::White:  return "#ffffff";
    }
    return "#ff4444";
}

/* =========================================================================
 * score_to_tier — same thresholds as spec §6 defaults
 * ========================================================================= */

Tier score_to_tier(double wf, double mc, double robustness) noexcept {
    double min_score = wf;
    if (mc         < min_score) min_score = mc;
    if (robustness < min_score) min_score = robustness;

    if (min_score >= 95.0) return Tier::White;
    if (min_score >= 90.0) return Tier::Green;
    if (min_score >= 80.0) return Tier::Yellow;
    if (min_score >= 70.0) return Tier::Orange;
    return Tier::Red;
}

/* =========================================================================
 * Size multiplier and flags per tier (§6 defaults)
 * ========================================================================= */

static double size_mult_for_tier(Tier t) noexcept {
    switch (t) {
        case Tier::Red:    return 0.0;
        case Tier::Orange: return 0.25;
        case Tier::Yellow: return 0.50;
        case Tier::Green:  return 1.00;
        case Tier::White:  return 1.50;
    }
    return 0.0;
}

static bool is_experimental(Tier t) noexcept {
    // orange and yellow are experimental
    return t == Tier::Orange || t == Tier::Yellow || t == Tier::Red;
}

static bool is_paper_only(Tier t) noexcept {
    // paper_only_below = yellow (spec default): orange and red
    return t == Tier::Orange || t == Tier::Red;
}

/* =========================================================================
 * build_tier_report — helper used by validator
 * ========================================================================= */

TierReport build_tier_report(
    double wf, double mc, double robustness,
    const std::string& manifest_name,
    const std::vector<StageResult>& stages
) {
    TierReport r;
    r.passed        = true;
    r.has_tier      = true;
    r.manifest_name = manifest_name;
    r.walk_forward  = wf;
    r.mc_bootstrap  = mc;
    r.robustness    = robustness;
    r.min_score     = std::min({wf, mc, robustness});
    r.tier          = score_to_tier(wf, mc, robustness);
    r.size_mult     = size_mult_for_tier(r.tier);
    r.experimental  = is_experimental(r.tier);
    r.paper_only    = is_paper_only(r.tier);
    r.stages        = stages;
    return r;
}

} /* namespace algoforge::algo_gen */
